### PR TITLE
Namecheap Instant Answer

### PR DESCRIPTION
**Pull request by tejasmanohar**
**What does your Instant Answer do?**
When a user's query begins with "namecheap", we do a domain lookup to the NC API. If the domain is available, a purchase link to Namecheap is provided. If not, a link to make an offer to the owner is provided via DomainAgents.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
The instant answer provides a simple way of purchasing domains straight from a user's DDG query.

**What is the data source for your Instant Answer? (Provide a link if possible)**
[NameCheap API](https://www.namecheap.com/support/api/intro.aspx)

**Why did you choose this data source?**
It provides responses for TLDs that relate to NameCheap. Modular + expandable :+1: 

**Are there any other alternative (better) data sources?**
Many alternative domain services with APIs... wouldn't say any are better but NameCheap has most relevant data

**What are some example queries that trigger this Instant Answer?**
`namecheap qwoenqwe.com`

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Domain enthusiasts, web power users, programmers, business owners

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No

**Which existing Instant Answers will this one supersede/overlap with?**
None

**Are you having any problems? Do you need our help with anything?**
We are using jQuery to parse XML into JSON. This works fine, but it may be better as a backend Perl service.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No costs. DuckDuckGo needs to acquire [NameCheap API](https://www.namecheap.com/support/api/intro.aspx) keys as instructed in documentation.

**Where did you hear about DuckDuckHack? (For first time contributors)**
Zaahir from DuckDuckGo

### Screenshots
https://imgur.com/I2MHif1,JxbJfEH#0

### CODE REVIEW

**Comment by moollaza** [comment]
Despite the silence, this PR is still alive -- speaking with Namecheap API team about a way to make the API request a little simpler.

**Comment by tejasmanohar** [comment]
@moollaza Teddy has updated the rate limits on the DuckDuckGo account. Let's get this deployed?

**Comment by jagtalon** [comment]
@moollaza @tejasmanohar I'm excited!

**Comment by moollaza** [comment]
@tejasmanohar alright sorry for the long delay. This is juuust about ready to go.

Can you please remove the console/debug code from the JS?

Also it seems that the API returns errors when you search for url with a subdomain -- let's just ignore those queries from the Perl, so we don't waste making an API call.

**Comment by tejasmanohar** [comment]
Resolved using a TLD regex. @moollaza

**Comment by moollaza** [comment]
@tejasmanohar wonderful, thanks a lot!

File: lib/DDG/Util/SpiceConstants.pm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package DDG::Util::SpiceConstants;
+
+use strict;
+use warnings;
+
+# regex for allowed TLDS (grabbed from DDG core repo, in /lib/DDG/Util/Constants.pm)
```
Comment (moollaza) [inline lib/DDG/Util/SpiceConstants.pm]: Out of curiosity, how'd you know the regex came from there?

Comment (tejasmanohar) [inline lib/DDG/Util/SpiceConstants.pm]: Original comment in `IsItUp.pm` 
said that. Didn't break into your servers for a regex, :stuck_out_tongue_winking_eye:

Comment (moollaza) [inline lib/DDG/Util/SpiceConstants.pm]: Hahaha I was looking for a comment but couldn't find anything!

Comment (tejasmanohar) [inline lib/DDG/Util/SpiceConstants.pm]: here it was! https://github.com/duckduckgo/zeroclickinfo-spice/commit/a75999c383e7a1a068f409af2d779221f762148a#diff-370547fe6407e842a2a7cc070b084393L25

**Comment by moollaza** [comment]
Alright this is ready for a design sign-off! @abeyang how's it looking?

https://ddh8.duckduckgo.com/?q=namecheap+duckduckgo.com&ia=namecheap
https://ddh8.duckduckgo.com/?q=namecheap+zaahir.me&ia=namecheap

------

I think we can probably remove the subtitle?

File: lib/DDG/Spice/Namecheap.pm [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+# environment variables:
+# DDG_SPICE_NAMECHEAP_USERNAME : Namecheap API username
+# DDG_SPICE_NAMECHEAP_APIKEY   : Namecheap API key
+# DDG_SPICE_NAMECHEAP_IP_ADDR  : Allowed end-User IP address
+my @query_list = (
+    ApiUser    => '{{ENV{DDG_SPICE_NAMECHEAP_USERNAME}}}', # Username required to access the API
+    ApiKey     => '{{ENV{DDG_SPICE_NAMECHEAP_APIKEY}}}',   # Password required used to access the API
+    UserName   => '{{ENV{DDG_SPICE_NAMECHEAP_USERNAME}}}', # same as ApiUser
+    Command    => 'namecheap.domains.check',               # API method
+    ClientIp   => '{{ENV{DDG_SPICE_NAMECHEAP_IP_ADDR}}}',  # End-user IP address
```
Comment (moollaza) [inline lib/DDG/Spice/Namecheap.pm]: @tejasmanohar can you change this to:

`ClientIp => '0.0.0.0'`

**Comment by abeyang** [comment]
Hey, looks good. I agree with @moollaza -- let's remove the subtitle as it is rather redundant.

**Comment by jagtalon** [comment]
@moollaza @abeyang should the tab name be "Domain" instead?

**Comment by moollaza** [comment]
@jagtalon ya that sounds good.

File: share/spice/namecheap/namecheap.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        var data = {
+            domainAvailable: available === "true",
+            domainName: domainName
+        };
+
+        Spice.add({
+            id               : 'namecheap',
+            name             : 'Namecheap',
```
Comment (moollaza) [inline share/spice/namecheap/namecheap.js]: We're trying to generalize the tab names and prefer not to show any brand names, (e.g. Amazon tab is called "products", Soundcloud is "music") so let's change this to `Domain`

**Comment by moollaza** [comment]
Seeing as the final fixes are quite minor I don't mind merging this and making the final adjustments myself.

@tejasmanohar thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by jagtalon** [comment]
@moollaza also the domain--we don't want bolding in the IA right? CC @abeyang 

screen shot 2015-03-04 at 3 43 25 pm

**Comment by tejasmanohar** [comment]
@moollaza @jagtalon thanks for merging, let me know if any other changes are to be done :smile:

**Comment by tejasmanohar** [comment]
Also, do you guys have an approximation on when this will be launched in production?

**Comment by moollaza** [comment]
@tejasmanohar about a week or so -- I'll be sure to ping you once it goes live

File: share/spice/namecheap/namecheap.js (hunk 3 above)
Comment (tejasmanohar) [inline share/spice/namecheap/namecheap.js]: +1

**Comment by moollaza** [comment]
@tejasmanohar congrats! It's live: https://duckduckgo.com/?q=namecheap+zpappis.com&ia=domain

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Pull request merged**